cli/src/formatted_diagnostic.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::HashMap;
+use std::fmt::{Debug, Display};
+use std::io;
+use std::ops::Add;
+
+use analyzer::diagnostic::ObservationTag;
+use analyzer::relations::SourceId;
+use ariadne::{Cache, Color, ColorGenerator, Config, Label, Report, ReportKind};
+use regex::Regex;
+use yansi::Paint;
+
+use context::source::Source;
+
+use crate::fix_ariadne_report;
+
+pub fn render_diagnostic<'a>(
+    diagnostic: analyzer::diagnostic::Diagnostic,
+    source_supplier: impl Fn(SourceId) -> Source<'a>,
+) -> io::Result<String> {
+    let mut colors = ColorGenerator::new();
+    let mut colormap = HashMap::new();
+    let mut sources = HashMap::new();
+
+    fn get_color(
+        colors: &mut ColorGenerator,
+        colormap: &mut HashMap<u8, Color>,
+        tag: Option<ObservationTag>,
+    ) -> Color {
+        match tag {
+            Some(ObservationTag::InFault) => Color::Red,
+            Some(ObservationTag::Declaration) => Color::Yellow,
+            Some(ObservationTag::Expected) => Color::Green,
+            Some(ObservationTag::Other(tag)) => {
+                *colormap.entry(tag).or_insert_with(|| colors.next())
+            }
+            None => colors.next(),
+        }
+    }
+
+    let main_source = source_supplier(diagnostic.source);
+    sources.insert(diagnostic.source, main_source);
+
+    let (code, color) = if diagnostic.identifier.critical() {
+        (
+            format!("error[E{:04}]", diagnostic.identifier.code()),
+            Color::Red,
+        )
+    } else {
+        (
+            format!("warn[W{:04}]", diagnostic.identifier.code()),
+            Color::Yellow,
+        )
+    };
+
+    let labels = diagnostic.observations.into_iter().map(|o| {
+        let source = o.foreign_env.unwrap_or(diagnostic.source);
```

Comment:
It seems like only observations have a source. Maybe remove the `SourceId` from `Diagnostic` and let `Observation` have a `pub source: SourceId` field.

Reply:
We would lose the origin of the diagnostic

Reply:
The diagnostic doesn't have the source, it is the definition, that is an `Observation`.

Reply:
We still lose the origin of the diagnostic